### PR TITLE
feat(passkeys): register

### PR DESCRIPTION
**Pull request by peintnermax**
### Definition of Ready

- [ ] I am happy with the code
- [ ] Short description of the feature/issue is added in the pr description
- [ ] PR is linked to the corresponding user story
- [ ] Acceptance criteria are met
- [ ] All open todos and follow ups are defined in a new ticket and justified
- [ ] Deviations from the acceptance criteria and design are agreed with the PO and documented.
- [ ] Jest unit tests ensure that components produce expected outputs on different inputs.
- [ ] Cypress integration tests ensure that login app pages work as expected. The ZITADEL API is mocked.
- [ ] No debug or dead code
- [ ] My code has no repetitions

### CODE REVIEW

File: apps/login/app/(login)/register/page.tsx [hunk 1]
```diff
@@ -3,24 +3,46 @@ import {
...
+import SetPasswordForm from "#/ui/SetPasswordForm";
+
+export default async function Page({
+  searchParams,
+}: {
+  searchParams: Record<string | number | symbol, string | undefined>;
+}) {
+  const { firstname, lastname, email } = searchParams;
+
+  const setPassword = !!(firstname && lastname && email);
```
Comment (eliobischof) [inline apps/login/app/(login)/register/page.tsx:16]: Shouldn't the user either select to set a password or to register w/o pw?
What's the idea behind these criteria?

File: apps/login/app/(login)/register/page.tsx [hunk 2]
```diff
@@ -3,24 +3,46 @@ import {
...
     server
   );
 
-  return (
+  return setPassword ? (
+    <div className="flex flex-col items-center space-y-4">
+      <h1>Set Password</h1>
+      <p className="ztdl-p">Set the password for your account</p>
+
+      {legal && passwordComplexitySettings && (
```
Comment (eliobischof) [inline apps/login/app/(login)/register/page.tsx:28]: Why do we need legal && passwordComplexitySettings to render the SetPasswordForm?
Can't we fall back to defaults?

Like this, we just render nothing if one of these is falsy, correct?

File: apps/login/app/(login)/signup/success/page.tsx [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+import Link from "next/link";
+
+type Props = {
+  searchParams: { [key: string]: string | string[] | undefined };
+};
+export default async function Page({ searchParams }: Props) {
+  return (
+    <div className="flex flex-col items-center space-y-4">
+      <h1>Registration successful</h1>
+      <p className="ztdl-p">You are registered.</p>
```
Comment (eliobischof) [inline apps/login/app/(login)/signup/success/page.tsx:11]: Should we check if that is true before rendering the page?
If not show an error page?

File: apps/login/app/(login)/signup/page.tsx [hunk 4]
```diff
@@ -0,0 +1,27 @@
+import {
+  getLegalAndSupportSettings,
+  getPasswordComplexitySettings,
+  server,
+} from "#/lib/zitadel";
+import RegisterForm from "#/ui/RegisterForm";
+
+export default async function Page() {
```
Comment (eliobischof) [inline apps/login/app/(login)/signup/page.tsx:8]: Is this page redundant? What's the difference between /register and /signup?

File: apps/login/app/passkeys/verify/route.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        passkeyId,
+        passkeyName,
+        publicKeyCredential,
+        userId
+      )
+        .then((resp) => {
+          return NextResponse.json(resp);
+        })
+        .catch((error) => {
+          return NextResponse.json(error, { status: 500 });
```
Comment (eliobischof) [inline apps/login/app/passkeys/verify/route.ts:38]: Can't it be a user error?

File: apps/login/app/passkeys/verify/route.ts [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        device.vendor || device.model ? ", " : ""
+      }${os.name}${os.name ? ", " : ""}${browser.name}`;
+    }
+    const sessionCookie = await getSessionCookieById(sessionId);
+
+    const session = await getSession(
+      server,
+      sessionCookie.id,
+      sessionCookie.token
+    );
```
Comment (eliobischof) [inline apps/login/app/passkeys/verify/route.ts:22]: Is this an unhandled promise rejection?

File: apps/login/app/session/route.ts [hunk 7]
```diff
@@ -94,6 +97,8 @@ export async function PUT(request: NextRequest) {
                   loginName: session.factors?.user?.loginName ?? "",
                 };
 
+                console.log("new token", recent.token, newCookie.token);
+
```
Comment (eliobischof) [inline apps/login/app/session/route.ts:101]:
```suggestion
```

File: apps/login/lib/demos.ts [hunk 8]
```diff
@@ -48,11 +48,11 @@ export const demos: { name: string; items: Item[] }[] = [
       //     slug: "passwordless",
       //     description: "The page to login a user with his passwordless device",
       //   },
-      //   {
-      //     name: "Passwordless Create",
-      //     slug: "passwordless/create",
-      //     description: "The page to add a users passwordless device",
-      //   },
+      {
```
Comment (eliobischof) [inline apps/login/lib/demos.ts:51]: Set Password is also available, right?
BTW, what's the difference between set and change pw?

File: apps/login/lib/zitadel.ts [hunk 9]
```diff
@@ -2,6 +2,7 @@ import {
   ZitadelServer,
   ZitadelServerOptions,
   user,
+  management,
```
Comment (eliobischof) [inline apps/login/lib/zitadel.ts:5]:
```suggestion
```

File: apps/login/lib/zitadel.ts [hunk 10]
```diff
@@ -181,4 +200,70 @@ export function setEmail(server: ZitadelServer, userId: string): Promise<any> {
   );
 }
 
+const bearerTokenMetadata = (token: string) =>
+  new Metadata({ authorization: `Bearer ${token}` });
```
Comment (eliobischof) [inline apps/login/lib/zitadel.ts:204]:
```suggestion
```

File: apps/login/lib/zitadel.ts [hunk 11]
```diff
@@ -20,7 +21,10 @@ import {
   VerifyEmailResponse,
   SetSessionResponse,
   DeleteSessionResponse,
+  VerifyPasskeyRegistrationResponse,
+  SetHumanPasswordResponse,
```
Comment (eliobischof) [inline apps/login/lib/zitadel.ts:25]:
```suggestion
```

File: apps/login/ui/Info.tsx [hunk 12]
```diff
@@ -0,0 +1,14 @@
+import { ExclamationTriangleIcon } from "@heroicons/react/24/outline";
```
Comment (eliobischof) [inline apps/login/ui/Info.tsx:1]: Is this not redundant with ui/Alert.tsx?

File: apps/login/ui/RegisterFormWithoutPassword.tsx [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+        lastName: values.lastname,
+      }),
+    });
+
+    if (!res.ok) {
+      setLoading(false);
+      throw new Error("Failed to register user");
+    }
+
+    setLoading(false);
```
Comment (eliobischof) [inline apps/login/ui/RegisterFormWithoutPassword.tsx:56]:
```suggestion
    setLoading(false);
    if (!res.ok) {
      throw new Error("Failed to register user");
    }
```

File: apps/login/ui/RegisterFormWithoutPassword.tsx [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+          createSessionWithLoginName(value.email).then(({ factors }) => {
+            return router.push(
+              `/passkey/add?` +
+                new URLSearchParams({ loginName: factors.user.loginName })
+            );
+          });
+        });
+    //   .then((resp: any) => {
+    //       return router.push(`/verify?userID=${resp.userId}`);
+    //     });
```
Comment (eliobischof) [inline apps/login/ui/RegisterFormWithoutPassword.tsx:95]:
```suggestion
```

**Review by eliobischof — CHANGES_REQUESTED** [review]
- Did you check against the ready for review criteria? For example, there are quite some console.logs and outcommented code. 
- If something needs to be changed, please check other occurrences. I didn't repeat the comments.
- I can't run it locally
  image

File: apps/login/app/(login)/signup/page.tsx (hunk 4 above)
Comment (peintnermax) [reply apps/login/app/(login)/signup/page.tsx:8]: one allows the user to set the auth method and one is the default register function which already shows the password

**Comment by peintnermax** [comment]
> * Did you check against the ready for review criteria? For example, there are quite some console.logs and outcommented code.
> * If something needs to be changed, please check other occurrences. I didn't repeat the comments.
> * I can't run it locally
>   image

I'll improve error handling such that errors like this don't show up and log the missing context..

**Comment by peintnermax** [comment]
> > * Did you check against the ready for review criteria? For example, there are quite some console.logs and outcommented code.
> > * If something needs to be changed, please check other occurrences. I didn't repeat the comments.
> > * I can't run it locally
> >   image
> 
> I'll improve error handling such that errors like this don't show up and log the missing context..

now errors are thrown and catched either as custom component if it makes sense, or the error boundary is shown with the message

https://github.com/zitadel/typescript/pull/29/commits/98fee5cb2bd3eb0bf38f99b0ba1ca4cec6e714ab
Bildschirmfoto 2023-06-21 um 13 15 14

File: apps/login/app/(login)/register/page.tsx (hunk 2 above)
Comment (peintnermax) [reply apps/login/app/(login)/register/page.tsx:28]: This is business logic. We can decide to have a fallback for the complexity settings, but legal is required to have urls for the privacy policy and ToS.. I decided to throw an error if one of the commands fails.

Comment (peintnermax) [reply apps/login/app/(login)/register/page.tsx:28]: if one of both are missing, an error is thrown and the error boundary is rendered.

File: apps/login/app/(login)/signup/success/page.tsx (hunk 3 above)
Comment (peintnermax) [reply apps/login/app/(login)/signup/success/page.tsx:11]: We don't show the page as we directly load `/accounts`. I've removed it.

File: apps/login/app/passkeys/verify/route.ts (hunk 5 above)
Comment (peintnermax) [reply apps/login/app/passkeys/verify/route.ts:38]: not directly.. once the browser is called the payload should be transmitted correctly. the name is generated from the user agent header if its not provided like in our case.

File: apps/login/app/passkeys/verify/route.ts (hunk 6 above)
Comment (peintnermax) [reply apps/login/app/passkeys/verify/route.ts:22]: if no session could be retrieved for the context a 500 is returned below.. Any ideas on how to improve this? 🤔

File: apps/login/lib/zitadel.ts (hunk 9 above)
Comment (peintnermax) [reply apps/login/lib/zitadel.ts:5]: done

File: apps/login/ui/Info.tsx (hunk 12 above)
Comment (peintnermax) [reply apps/login/ui/Info.tsx:1]: yes. removed

File: apps/login/app/(login)/register/page.tsx (hunk 1 above)
Comment (eliobischof) [reply apps/login/app/(login)/register/page.tsx:16]: thanks for clarifying :+1:

File: apps/login/app/(login)/signup/page.tsx (hunk 4 above)
Comment (eliobischof) [reply apps/login/app/(login)/signup/page.tsx:8]: I don't get why we need a default register function which already shows the password. If we need a preselection, why not preselect the passkey? For me, the difference between the terms signup and register is obscure.

File: apps/login/app/passkeys/verify/route.ts (hunk 6 above)
Comment (eliobischof) [reply apps/login/app/passkeys/verify/route.ts:22]: something like catching the error and returning this specific error to the user. But we anyway have another issue for this...

File: apps/login/lib/demos.ts (hunk 8 above)
Comment (eliobischof) [reply apps/login/lib/demos.ts:26]: > but the idea is to have an endpoint where I can change my password with an active session

And what does the other endpoint? Just for my understanding. I also didn't find the answer in the epics issues.

File: apps/login/app/(login)/signup/page.tsx (hunk 4 above)
Comment (peintnermax) [reply apps/login/app/(login)/signup/page.tsx:8]: This is just because I kind of wanted to deprecated the old functionallity

File: apps/login/lib/demos.ts (hunk 8 above)
Comment (peintnermax) [reply apps/login/lib/demos.ts:26]: which one? password?

Comment (eliobischof) [reply apps/login/lib/demos.ts:26]: is "reset" mean with "set", so the user has to change it?

File: acceptance/docker-compose.yaml
Comment (eliobischof) [inline acceptance/docker-compose.yaml:1]: Please either enforce yaml linting or undo these changes

File: apps/login/app/(login)/passkey/add/page.tsx [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+          displayName={sessionFactors.factors?.user?.displayName}
+          showDropdown
+        ></UserAvatar>
+      )}
+      <p className="ztdl-p mb-6 block">{description}</p>
+
+      <Alert type={AlertType.INFO}>
+        <span>
+          A passkey is an authentication method on a device like your
+          fingerprint, Apple FaceID or similar.{" "}
```
Comment (eliobischof) [inline apps/login/app/(login)/passkey/add/page.tsx:47]: What is `{" "}`?

File: apps/login/app/passkeys/route.ts [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+    if (userId) {
+      return createPasskeyRegistrationLink(userId)
+        .then((resp) => {
+          const code = resp.code;
+          return registerPasskey(userId, code, domain).then((resp) => {
+            return NextResponse.json(resp);
+          });
+        })
+        .catch((error) => {
+          console.error("error on creating passkey registration link");
```
Comment (eliobischof) [inline apps/login/app/passkeys/route.ts:36]: I think we don't want to log to stdout, right?

File: apps/login/lib/zitadel.ts [hunk 17]
```diff
@@ -127,30 +135,36 @@ export type AddHumanUserData = {
...
+            username: email,
+            profile: { firstName, lastName },
+            password: { password },
+          }
+        : {
+            email: { email },
+            username: email,
+            profile: { firstName, lastName },
+          },
       {}
```
Comment (eliobischof) [inline apps/login/lib/zitadel.ts:160]: let's use the spread operator to make it more DRY

File: apps/login/ui/RegisterPasskey.tsx [hunk 18]
```diff
@@ -0,0 +1,207 @@
...
+          })
+          .catch((error) => {
+            console.error(error);
+            setLoading(false);
+            //   setError(error);
+
+            return null;
+          });
+      }
+      //   return router.push(`/accounts`);
```
Comment (eliobischof) [inline apps/login/ui/RegisterPasskey.tsx:160]: Please incomment or delete outcommented changes

File: apps/login/ui/UserAvatar.tsx [hunk 19]
```diff
@@ -14,20 +14,20 @@ export default function UserAvatar({
...
   return (
-    <div className="flex h-full w-full flex-row items-center rounded-full border p-[1px] dark:border-white/20">
+    <div className="flex h-full flex-row items-center rounded-full border p-[1px] dark:border-white/20">
       <div>
         <Avatar
           size="small"
-          name={displayName ?? loginName}
-          loginName={loginName}
+          name={displayName ?? loginName ?? "A"}
+          loginName={loginName ?? "A"}
```
Comment (eliobischof) [inline apps/login/ui/UserAvatar.tsx:22]: What about `?` instead of `A`?

**Review by eliobischof — COMMENTED** [review]
I stll have this error
image

File: apps/login/ui/UserAvatar.tsx (hunk 19 above)
Comment (peintnermax) [reply apps/login/ui/UserAvatar.tsx:22]: ? looks like an error to me.. I check if it can be rendered without a char

**Comment by peintnermax** [comment]
> I stll have this error image

That because you are not allowed to go to passkey/add without an active session

File: apps/login/ui/RegisterPasskey.tsx (hunk 18 above)
Comment (peintnermax) [reply apps/login/ui/RegisterPasskey.tsx]: done

File: apps/login/app/(login)/passkey/add/page.tsx (hunk 15 above)
Comment (peintnermax) [reply apps/login/app/(login)/passkey/add/page.tsx:47]: removed

File: apps/login/lib/zitadel.ts (hunk 17 above)
Comment (peintnermax) [reply apps/login/lib/zitadel.ts:160]: done

File: apps/login/ui/UserAvatar.tsx (hunk 19 above)
Comment (peintnermax) [reply apps/login/ui/UserAvatar.tsx]: removed the label entirely with emptystring

File: apps/login/app/passkeys/route.ts (hunk 16 above)
Comment (peintnermax) [reply apps/login/app/passkeys/route.ts:36]: i left it too, otherwise you dont see it in the browser console

Comment (eliobischof) [reply apps/login/app/passkeys/route.ts:36]: These are basically access logs, that we should be able to query from the LB, right? Otherwise we should configure it globally, IMO, with an interceptor or so.

Comment (peintnermax) [reply apps/login/app/passkeys/route.ts:36]: yes

**Review by eliobischof — APPROVED** [review]
